LibreHardwareMonitorLib/Hardware/Battery/BatteryGroup.cs

```diff
@@ -0,0 +1,231 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Text;
+using Microsoft.Win32.SafeHandles;
+using LibreHardwareMonitor.Interop;
+
+namespace LibreHardwareMonitor.Hardware.Battery
+{
+    internal struct BatteryReportInfo
+    {
+        public string Name { get; set; }
+        public uint Tag { get; set; }
+        public SafeFileHandle BatteryHandle { get; set; }
+        public Kernel32.BATTERY_INFORMATION MoreInformation { get; set; }
+    }
+
+    internal class BatteryGroup : IGroup
+    {
+        private List<Hardware> _hardware = new();
+        private List<BatteryReportInfo> _batteriesToReport = new();
+
+        public IReadOnlyList<IHardware> Hardware => _hardware;
+
+        private static IntPtr Offset(IntPtr pointer, long offset) => new IntPtr(pointer.ToInt64() + offset);
+        private static int GetCharSize(CharSet charSet = CharSet.Auto) => charSet == CharSet.Auto ? Marshal.SystemDefaultCharSize : (charSet == CharSet.Unicode ? 2 : 1);
+        private static bool IsValue(IntPtr ptr) => ptr.ToInt64() >> 16 == 0;
+
+        private static string GetString(IntPtr ptr, CharSet charSet = CharSet.Auto, long allocatedBytes = long.MaxValue)
+        {
+            if (IsValue(ptr)) return null;
+            var sb = new StringBuilder();
+            unsafe
+            {
+                var chkLen = 0L;
+                if (GetCharSize(charSet) == 1)
+                {
+                    for (var uptr = (byte*)ptr; chkLen < allocatedBytes && *uptr != 0; chkLen++, uptr++)
+                        sb.Append((char)*uptr);
+                }
+                else
+                {
+                    for (var uptr = (ushort*)ptr; chkLen + 2 <= allocatedBytes && *uptr != 0; chkLen += 2, uptr++)
+                        sb.Append((char)*uptr);
+                }
+            }
+            return sb.ToString();
+        }
+
+        public BatteryGroup(ISettings settings)
+        {
+            // No implementation for battery information on Unix systems
+            if (Software.OperatingSystem.IsUnix)
+            {
+                return;
+            }
+
+            IntPtr hdev = SetupApi.SetupDiGetClassDevs(ref SetupApi.GUID_DEVICE_BATTERY, IntPtr.Zero, IntPtr.Zero, SetupApi.DIGCF_PRESENT | SetupApi.DIGCF_DEVICEINTERFACE);
+            if (SetupApi.INVALID_HANDLE_VALUE != hdev)
+            {
+                // Limit battery search to 100 batteries
+                for (uint idev = 0; idev < 100; idev++)
```

Comment:
While I seriously doubt there would be more than 100 batteries, this should just loop until ERROR_NO_MORE_ITEMS is the last error.

Reply:
Yes, you're right. That's from the C++ code I've used as a reference. It's fixed now.